File: src/main/java/org/nomad/pithos/components/GroupLedger.java
package org.nomad.pithos.components;

import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
import it.unimi.dsi.fastutil.objects.ObjectArrayList;
import it.unimi.dsi.fastutil.objects.ObjectList;
import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
import org.nomad.grpc.superpeerservice.MultiMapPair;
import org.nomad.pithos.mappers.CustomMappers;
import org.nomad.pithos.models.MetaData;

import java.util.Collection;
import java.util.stream.Collectors;

public class GroupLedger implements GenericGroupLedger {
    private static GroupLedger INSTANCE;
    private final Ledger objectLedger = new Ledger();
    private final Ledger peerLedger = new Ledger();

    public synchronized static GroupLedger getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new GroupLedger();
        }
        return INSTANCE;
    }

    /* Repair Data
    ================================================================================================================*/
    @Override
    public int countReplicas(String objectId) {
        return objectLedger.countReplicas(objectId);
    }

    @Override
    public Object2ObjectOpenHashMap<String, Integer> objectsThatNeedRepair(int rf) {
        cleanExpiredObjects();
        Object2ObjectOpenHashMap<String, Integer> objectRepairMap = new Object2ObjectOpenHashMap<>();
        ObjectList<String> objectsNeedingRepair = objectLedger.getObjectsNeedingRepair(rf);
        objectsNeedingRepair.forEach(key -> {
            int count = countReplicas(key);
            if (count < rf) {
                objectRepairMap.put(key, (rf - count));
            }
        });
        return objectRepairMap;
    }

    @Override
    public ObjectList<String> removePeersStoringObject(Collection<String> peers, String objectId) {
        return new ObjectArrayList<>(peers.stream().filter(peerId -> !peerLedger.containsEntryNoTTL(peerId, objectId)).collect(Collectors.toList()));
    }

    @Override
    public ObjectList<String> removePeersNotStoringObject(Collection<String> peers, String objectId) {
        return new ObjectArrayList<>(peers.stream().filter(peerId -> peerLedger.containsEntryNoTTL(peerId, objectId)).collect(Collectors.toList()));
    }


    /* Adding Data
    ================================================================================================================*/

    @Override
    public void populateGroupLedger(MultiMapPair objectLedger, MultiMapPair peerLedger) {
        this.objectLedger.populateLedger(CustomMappers.INSTANCE.convertToMap(objectLedger));
        this.peerLedger.populateLedger(CustomMappers.INSTANCE.convertToMap(peerLedger));
    }

    @Override
    public boolean addToObjectLedger(String objectId, MetaData peerId) {
        return objectLedger.add(objectId, peerId);
    }

    @Override
    public boolean addToPeerLedger(String peerId, MetaData objectId) {
        return peerLedger.add(peerId, objectId);
    }

    @Override
    public boolean objectLedgerContainsKey(String key) {
        return objectLedger.containsKey(key);
    }

    @Override
    public boolean peerLedgerContainsKey(String key) {
        return peerLedger.containsKey(key);
    }

    @Override
    public boolean objectLedgerContainsEntry(String key, MetaData value) {
        return objectLedger.containsEntry(key, value);
    }

    @Override
    public boolean peerLedgerContainsEntry(String key, MetaData value) {
        return peerLedger.containsEntry(key, value);
    }

    @Override
    public boolean thisPeerContainsObject(String key, String value) {
        return peerLedger.containsEntry(key, value);
    }

    /* Removing Data
    ================================================================================================================*/

    @Override
    public void removeFromObjectLedger(String objectId, MetaData peerId) {
        objectLedger.remove(objectId, peerId);
    }

    @Override
    public void removeFromPeerLedger(String peerId, MetaData objectId) {
        peerLedger.remove(peerId, objectId);
    }

    @Override
    public void removePeerFromGroupLedger(String peerId) {
        peerLedger.removeAll(peerId);
        objectLedger.removeValueNoTTL(peerId);
    }

    @Override
    public void removeObjectFromGroupLedger(String objectId) {
        objectLedger.removeAll(objectId);
        peerLedger.removeValueNoTTL(objectId);
    }

    @Override
    public boolean cleanExpiredObjects() {
        boolean peerLedgerCleaned = peerLedger.cleanExpiredObjects();
        boolean objectLedgerCleaned = objectLedger.cleanExpiredObjects();
        return peerLedgerCleaned && objectLedgerCleaned;
    }

    @Override
    public void clearAll() {
        peerLedger.clear();
        objectLedger.clear();
    }

    /* Viewing Data
    ================================================================================================================*/

    @Override
    public ObjectOpenHashSet<String> getAllGroupObjects() {
        return new ObjectOpenHashSet<>(objectLedger.keySet());
    }

    /**
     * Purely a method used for manual testing
     */
    @Override
    public void printObjectLedger() {
//        System.out.println("Object ledger:\n" + objectLedger.ledgerMultiMap.toString());
    }

    /**
     * Purely a method used for manual testing
     */
    @Override
    public void printPeerLedger() {
//        System.out.println("Peer ledger:\n" + peerLedger.ledgerMultiMap.toString());
    }

    @Override
    public Object2ObjectOpenHashMap<String, Collection<MetaData>> getObjectLedgerAsMap() {
        return new Object2ObjectOpenHashMap<>(objectLedger.asMap());
    }

    @Override
    public Object2ObjectOpenHashMap<String, Collection<MetaData>> getPeerLedgerAsMap() {
        return new Object2ObjectOpenHashMap<>(peerLedger.asMap());
    }

    @Override
    public Object2ObjectOpenHashMap<String, org.nomad.grpc.superpeerservice.MetaDataCollection> getGrpcObjectLedger() {
        return new Object2ObjectOpenHashMap<>(objectLedger.asGrpcMetaDataCollection());
    }

    @Override
    public Object2ObjectOpenHashMap<String, org.nomad.grpc.superpeerservice.MetaDataCollection> getGrpcPeerLedger() {
        return new Object2ObjectOpenHashMap<>(peerLedger.asGrpcMetaDataCollection());
    }

    /* Getters & Setters
    ================================================================================================================*/

    @Override
    public Ledger getObjectLedger() {
        return objectLedger;
    }

    @Override
    public Ledger getPeerLedger() {
        return peerLedger;
    }

    @Override
    public Ledger setObjectLedger(Object2ObjectOpenHashMap<String, Collection<String>> map) {
        return objectLedger;
    }

    @Override
    public Ledger setPeerLedger() {
        return peerLedger;
    }
}

File: src/main/java/org/nomad/pithos/components/GroupLedger.kt
package org.nomad.pithos.components

import org.nomad.pithos.mappers.CustomMappers.convertToMap
import org.nomad.pithos.components.GenericGroupLedger
import org.nomad.pithos.components.Ledger
import java.util.stream.Collectors
import org.nomad.grpc.superpeerservice.MultiMapPair
import org.nomad.pithos.mappers.CustomMappers
import org.nomad.grpc.superpeerservice.MetaDataCollection
import org.nomad.pithos.components.GroupLedger
import org.nomad.pithos.models.MetaData

class GroupLedger : GenericGroupLedger {
    /* Getters & Setters
    ================================================================================================================*/  override val objectLedger = Ledger()
    override val peerLedger = Ledger()

    /* Repair Data
      ================================================================================================================*/
    override fun countReplicas(objectId: String?): Int {
        return objectLedger.countReplicas(objectId)
    }

    override fun objectsThatNeedRepair(rf: Int): HashMap<String, Int> {
        cleanExpiredObjects()
        val objectRepairMap: HashMap<String, Int> = HashMap()
        val objectsNeedingRepair: ArrayList<String> = objectLedger.getObjectsNeedingRepair(rf)
        objectsNeedingRepair.forEach { key ->
            val count = countReplicas(key)
            if (count < rf) {
                objectRepairMap.put(key, rf - count)
            }
        }
        return objectRepairMap
    }

    override fun removePeersStoringObject(peers: Collection<String?>?, objectId: String?): ArrayList<String?>? {
        return ArrayList(peers!!.stream().filter { peerId: String? -> !peerLedger.containsEntryNoTTL(peerId, objectId) }.collect(Collectors.toList<Any?>()))
    }

    override fun removePeersNotStoringObject(peers: Collection<String?>?, objectId: String?): ArrayList<String?>? {
        return ArrayList(peers!!.stream().filter { peerId: String? -> peerLedger.containsEntryNoTTL(peerId, objectId) }.collect(Collectors.toList<Any?>()))
    }

    /* Adding Data
    ================================================================================================================*/
    override fun populateGroupLedger(objectLedger: MultiMapPair?, peerLedger: MultiMapPair?) {
        this.objectLedger.populateLedger(CustomMappers.INSTANCE.convertToMap(objectLedger!!))
        this.peerLedger.populateLedger(CustomMappers.INSTANCE.convertToMap(peerLedger!!))
    }

    override fun addToObjectLedger(objectId: String?, peerId: MetaData?): Boolean {
        return objectLedger.add(objectId, peerId)
    }

    override fun addToPeerLedger(peerId: String?, objectId: MetaData?): Boolean {
        return peerLedger.add(peerId, objectId)
    }

    override fun objectLedgerContainsKey(key: String?): Boolean {
        return objectLedger.containsKey(key)
    }

    override fun peerLedgerContainsKey(key: String?): Boolean {
        return peerLedger.containsKey(key)
    }

    override fun objectLedgerContainsEntry(key: String?, value: MetaData?): Boolean {
        return objectLedger.containsEntry(key, value)
    }

    override fun peerLedgerContainsEntry(key: String?, value: MetaData?): Boolean {
        return peerLedger.containsEntry(key, value)
    }

    override fun thisPeerContainsObject(key: String?, value: String?): Boolean {
        return peerLedger.containsEntry(key, value)
    }

    /* Removing Data
    ================================================================================================================*/
    override fun removeFromObjectLedger(objectId: String?, peerId: MetaData?) {
        objectLedger.remove(objectId, peerId)
    }

    override fun removeFromPeerLedger(peerId: String?, objectId: MetaData?) {
        peerLedger.remove(peerId, objectId)
    }

    override fun removePeerFromGroupLedger(peerId: String?) {
        peerLedger.removeAll(peerId)
        objectLedger.removeValueNoTTL(peerId)
    }

    override fun removeObjectFromGroupLedger(objectId: String?) {
        objectLedger.removeAll(objectId)
        peerLedger.removeValueNoTTL(objectId)
    }

    override fun cleanExpiredObjects(): Boolean {
        val peerLedgerCleaned = peerLedger.cleanExpiredObjects()
        val objectLedgerCleaned = objectLedger.cleanExpiredObjects()
        return peerLedgerCleaned && objectLedgerCleaned
    }

    override fun clearAll() {
        peerLedger.clear()
        objectLedger.clear()
    }

    /* Viewing Data
    ================================================================================================================*/
    override val allGroupObjects: Set<String>
        get() = HashSet() (objectLedger.keySet())

    /**
     * Purely a method used for manual testing
     */
    override fun printObjectLedger() {
//        System.out.println("Object ledger:\n" + objectLedger.ledgerMultiMap.toString());
    }

    /**
     * Purely a method used for manual testing
     */
    override fun printPeerLedger() {
//        System.out.println("Peer ledger:\n" + peerLedger.ledgerMultiMap.toString());
    }

    override val objectLedgerAsMap: HashMap<String, Collection<MetaData>>
        get() = HashMap(objectLedger.asMap())
    override val peerLedgerAsMap: HashMap<String, Collection<MetaData>>
        get() = HashMap(peerLedger.asMap())
    override val grpcObjectLedger: HashMap<String, MetaDataCollection>
        get() = HashMap(objectLedger.asGrpcMetaDataCollection())
    override val grpcPeerLedger: HashMap<String, MetaDataCollection>
        get() = HashMap(peerLedger.asGrpcMetaDataCollection())

    override fun setObjectLedger(map: HashMap<String?, Collection<String?>?>?): Ledger {
        return objectLedger
    }

    override fun setPeerLedger(): Ledger? {
        return peerLedger
    }

    companion object {
        private var INSTANCE: GroupLedger? = null

        @JvmStatic
        @get:Synchronized
        val instance: GroupLedger?
            get() {
                if (INSTANCE == null) {
                    INSTANCE = GroupLedger()
                }
                return INSTANCE
            }
    }
}